src/lib/io/sol-ipm.c

```diff
@@ -0,0 +1,370 @@
+                message->base.mem,
+                message->base.size);
+
+            remote_blob->remote = &message->base;
+            receiver->receive_cb((void *)receiver->data, message->id,
+                &remote_blob->base);
+            return;
+        }
+    }
+
+    SOL_INF("Processed IPM id %" PRIu32 " but no receiver was found!",
+        message->id);
+
+fail:
+    /* Let's unref this blob on remote */
+    free_remote_blob(&message->base);
+}
+
+static void
+ipm_reaper_process(void *data)
+{
+    int idx;
+    struct sol_blob_ipm_sent *message = data;
+    struct sol_ipm_consumed_handler *handler;
+
+    SOL_VECTOR_FOREACH_IDX (&ipm_consumed_handlers, handler, idx) {
+        if (handler->id == message->id) {
+            handler->message_consumed_cb((void *)handler->data, message->id,
+                message->base.parent);
+            sol_blob_unref(&message->base);
+            free(message); /* This blob is SOL_BLOB_TYPE_NO_FREE, so we need to
+                              free it after using it. Parent takes care of freeing
+                              blob->mem*/
+            return;
+        }
+    }
+}
+
+/* Note: this function runs in interrupt context */
+static void
+ipm_receiver_isr_cb(void *context, uint32_t id, volatile void *data)
+{
+    struct sol_blob_ipm_sent *remote_blob = (struct sol_blob_ipm_sent *)AS_PTR(data);
+    struct mainloop_event me;
+
+    /* id == 0 messages. Those messages signal sender
+    * core that blob was consumed and may be freed */
+    if (id == SOL_IPM_ID_FREE_REMOTE) {
+        me.cb = ipm_reaper_process;
+        me.data = remote_blob;
+        sol_mainloop_event_post(&me);
+        return;
+    }
+
+    me.cb = ipm_receiver_process,
+    me.data = remote_blob;
+    sol_mainloop_event_post(&me);
+}
+
+static int
+init(void)
+{
+    int r;
+    uint32_t max_data_size;
+
+    ipm_receiver = device_get_binding("message_ipm_receiver");
+    SOL_NULL_CHECK_MSG(ipm_receiver, -EINVAL, "Could not get IPM receiver channel");
+
+    ipm_sender = device_get_binding("message_ipm_sender");
+    SOL_NULL_CHECK_MSG(ipm_sender, -EINVAL, "Could not get IPM sender channel");
+
+    max_data_size = ipm_max_data_size_get(ipm_sender);
+    if (max_data_size < sizeof(void *)) {
+        SOL_WRN("IPM max data size < sizeof(void *)");
+        return -EINVAL;
+    }
+
+    ipm_max_id = ipm_max_id_val_get(ipm_sender);
+
+    ipm_register_callback(ipm_receiver, ipm_receiver_isr_cb, NULL);
+    r = ipm_set_enabled(ipm_receiver, 1);
+    if (r != 0) {
+        SOL_WRN("Could not enable IPM receiver: %s", sol_util_strerrora(r));
+        return -r;
+    }
+
+    initialised = true;
+
+    return 0;
+}
+
+SOL_API int
+sol_ipm_set_receiver(uint32_t id,
+    void (*receive_cb)(void *cb_data, uint32_t id, struct sol_blob *message),
+    const void *data)
+{
+    int idx;
+    struct sol_ipm_receiver_entry *receiver;
+
+    if (!initialised) {
+        int r = init();
+        SOL_INT_CHECK(r, != 0, r);
```

Comment:
also, shouldn't you initialize only in the success case where you need to `sol_vector_append()`?  Otherwise on errors, you did that for no good reason.